crates/mun_compiler/src/diagnostics_snippets.rs

```diff
@@ -0,0 +1,372 @@
+}
+
+fn syntax_node_ptr_location(
+    syntax_node_ptr: SyntaxNodePtr,
+    parse: &Parse<SourceFile>,
+) -> TextRange {
+    match syntax_node_ptr.kind() {
+        SyntaxKind::FUNCTION_DEF => {
+            ast::FunctionDef::cast(syntax_node_ptr.to_node(parse.tree().syntax()))
+                .map(|f| f.signature_range())
+                .unwrap_or_else(|| syntax_node_ptr.range())
+        }
+        SyntaxKind::STRUCT_DEF => {
+            ast::StructDef::cast(syntax_node_ptr.to_node(parse.tree().syntax()))
+                .map(|s| s.signature_range())
+                .unwrap_or_else(|| syntax_node_ptr.range())
+        }
+        _ => syntax_node_ptr.range(),
+    }
+}
+
+pub(crate) fn syntax_error(
+    syntax_error: &SyntaxError,
+    _: &impl HirDatabase,
+    _: &Parse<SourceFile>,
+    relative_file_path: &str,
+    source_code: &str,
+    source_code_len: usize,
+    line_index: &Arc<LineIndex>,
+) -> Snippet {
+    let mut snippet = SnippetBuilder::new()
+        .title(
+            AnnotationBuilder::new(AnnotationType::Error)
+                .label("syntax error".to_string())
+                .build(),
+        )
+        .slice(
+            SliceBuilder::new(true)
+                .origin(relative_file_path.to_string())
+                .source_annotation(
+                    (
+                        syntax_error.location().offset().to_usize(),
+                        syntax_error.location().end_offset().to_usize(),
+                    ),
+                    syntax_error.to_string(),
+                    AnnotationType::Error,
+                )
+                .build(&source_code, source_code_len, &line_index),
+        )
+        .build();
+    // Add one to right range to make highlighting range here visible on output
+    snippet.slices[0].annotations[0].range.1 += 1;
+
+    snippet
+}
+
+pub(crate) fn generic_error(
+    diagnostic: &dyn HirDiagnostic,
+    _: &impl HirDatabase,
+    _: &Parse<SourceFile>,
+    relative_file_path: &str,
+    source_code: &str,
+    source_code_len: usize,
+    line_index: &Arc<LineIndex>,
+) -> Snippet {
+    SnippetBuilder::new()
+        .title(
+            AnnotationBuilder::new(AnnotationType::Error)
+                .label(diagnostic.message())
+                .build(),
+        )
+        .slice(
+            SliceBuilder::new(true)
+                .origin(relative_file_path.to_string())
+                .source_annotation(
+                    text_range_to_tuple(diagnostic.highlight_range()),
+                    diagnostic.message(),
+                    AnnotationType::Error,
+                )
+                .build(&source_code, source_code_len, &line_index),
+        )
+        .build()
+}
+
+pub(crate) fn unresolved_value_error(
+    diagnostic: &mun_hir::diagnostics::UnresolvedValue,
+    _: &impl HirDatabase,
+    parse: &Parse<SourceFile>,
+    relative_file_path: &str,
+    source_code: &str,
+    source_code_len: usize,
```

Comment:
A slice includes the length, so you should be able to incorporate this in `source_code: &str`